105-strimzi-drain-cleaner-namespace-filtering-proposal.md

```diff
@@ -0,0 +1,269 @@
+
+**Example 2: Unwatched Namespace**
+- Eviction request for pod in `app-namespace` → Immediate allow
+- Log: `DEBUG: Ignoring eviction request for Pod some-pod in namespace app-namespace - namespace not in watch list`
+- Result: Eviction proceeds normally, no API calls made
+
+**Example 3: Default Behavior**
+```yaml
+env:
+  - name: STRIMZI_DRAIN_WATCH_NAMESPACES
+    value: ""
+```
+- All namespaces processed (current behavior)
+- Full backward compatibility maintained
+
+### Security Considerations
+
+1. **No RBAC Changes Required**: Existing cluster-wide permissions remain necessary for webhook operation
+2. **Application-Level Security**: Filtering prevents unnecessary API calls to restricted namespaces
+3. **Audit Trail**: All filtering decisions logged at appropriate levels
+4. **Fail-Safe Design**: Invalid configuration defaults to watching all namespaces
+
+## Alternatives Considered
+
+### 1. RBAC-Level Filtering
+**Approach**: Use Kubernetes RBAC to restrict access
+**Rejected Because**: 
+- Webhooks require cluster-wide registration
+- Cannot selectively grant namespace access for webhook operations
+- Would break fundamental webhook architecture
+
+### 2. Multiple Drain Cleaner Instances
+**Approach**: Deploy separate drain cleaners per namespace
+**Rejected Because**:
+- Requires multiple ValidatingWebhookConfigurations
+- Potential conflicts and performance issues
+- Goes against "one per cluster" design principle
+- Complex certificate and configuration management
+
+### 3. Webhook Namespace Selectors
+**Approach**: Use ValidatingWebhookConfiguration `namespaceSelector`
+**Rejected Because**:
+- Requires cluster-level configuration changes
+- Less flexible than application-level configuration
+- Cannot be easily managed through Helm values
+- Harder to update without webhook re-registration
+
+### 4. Admission Controller Scope Changes
+**Approach**: Modify webhook registration scope
+**Rejected Because**:
+- Changes fundamental architecture
+- Reduces drain cleaner effectiveness
+- Complex migration path for existing deployments
+
+## Test Plan
+
+### Unit Tests
+- `testNamespaceFilteringWithWatchedNamespace()`: Verify processing of watched namespaces
+- `testNamespaceFilteringWithUnwatchedNamespace()`: Verify skipping of unwatched namespaces  
+- `testNamespaceFilteringWithEmptyConfiguration()`: Verify backward compatibility
+- `testNamespaceFilteringWithWhitespaceHandling()`: Verify configuration parsing
+
+### Integration Tests
+- Deploy drain cleaner with namespace filtering in multi-namespace test environment
+- Verify eviction handling for both watched and unwatched namespaces
+- Confirm no API calls made to unwatched namespaces
+- Validate log output and behavior
+
+### Backward Compatibility Tests
+- Verify existing deployments continue working without configuration changes
+- Test upgrade scenarios from previous versions
+- Confirm default behavior matches current implementation
```

Comment:
I guess backward compatibility will be covered by the existing tests. So it does not need anything special. We also do not have any upgrade tests for Drain Cleaner, as it is just a drop-in replacement of a stateless application.